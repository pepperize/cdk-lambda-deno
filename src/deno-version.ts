// https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
const regex =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/g;

export class DenoVersion {
  /**
   * 1.23.3
   */
  public static readonly V1_23_3 = new DenoVersion("1.23.3");

  /**
   * Specify a custom version.
   * Use this if the version you need is not available in one of the predefined versions.
   *
   * @param version The semantic version number without prefix 'v'.
   */
  public static of(version: string): DenoVersion {
    if (!regex.test(version)) {
      throw new Error(`'${version}' is not a valid semantic version`);
    }

    return new DenoVersion(version);
  }

  /**
   * The version string.
   */
  public readonly version: string;

  private constructor(version: string) {
    this.version = version;
  }
}
