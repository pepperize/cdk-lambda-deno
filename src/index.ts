export * from "./deno-function";
export * from "./deno-layer";
export * from "./deno-version";
