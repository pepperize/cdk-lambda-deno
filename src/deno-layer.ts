import * as path from "path";
import { DockerImage, Stack } from "aws-cdk-lib";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import { DenoVersion } from "./deno-version";

export interface DenoLayerOptions {
  /**
   * @default 1.23.3
   */
  readonly version?: DenoVersion;
}

export class DenoLayer extends lambda.LayerVersion {
  public static getOrCreate(scope: Construct, options: DenoLayerOptions): DenoLayer {
    const stack = Stack.of(scope);
    const id = "DenoLayer";
    const existing = stack.node.tryFindChild(id);
    return (existing as DenoLayer) || new DenoLayer(stack, id, options);
  }

  public constructor(scope: Construct, id: string, options: DenoLayerOptions) {
    const version = options.version ?? DenoVersion.V1_23_3;

    const image = DockerImage.fromBuild(path.join(__dirname, "../layer"), {
      buildArgs: { DENO_VERSION: version.version },
    });
    image.cp("/layer.zip", path.join(__dirname));

    const props: lambda.LayerVersionProps = {
      code: lambda.Code.fromAsset(path.join(__dirname, "layer.zip")),
      description: "/bin/deno",
    };

    super(scope, id, props);
  }
}
