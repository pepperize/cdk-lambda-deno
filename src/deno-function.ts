import * as lambda from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";
import { DenoLayer } from "./deno-layer";
import { DenoVersion } from "./deno-version";

export interface DenoFunctionProps extends lambda.FunctionOptions {
  readonly code: lambda.Code;
  readonly handler?: string;
  readonly version?: DenoVersion;
}

export class DenoFunction extends lambda.Function {
  public constructor(scope: Construct, id: string, props: DenoFunctionProps) {
    const layer: lambda.ILayerVersion = DenoLayer.getOrCreate(scope, { version: props.version });
    const layers = props.layers || [];
    layers.unshift(layer);

    const handler = props.handler ?? "index.handler";

    super(scope, id, {
      runtime: lambda.Runtime.PROVIDED_AL2,
      handler: handler,
      ...props,
      layers: layers,
    });
  }
}
