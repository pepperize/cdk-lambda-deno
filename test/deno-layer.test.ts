import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { DenoLayer } from "../src";

describe("DenoLayer", () => {
  it("should match snapshot", () => {
    // Given
    const stack = new Stack();

    // When
    DenoLayer.getOrCreate(stack);

    // Then
    const template = Template.fromStack(stack);

    expect(template.toJSON()).toMatchSnapshot();
  });
  it("should synthesize to a layer version", () => {
    // Given
    const stack = new Stack();

    // When
    new DenoLayer(stack, "Layer");

    // Then
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::LayerVersion", {
      Description: "/bin/deno",
    });
  });
});
