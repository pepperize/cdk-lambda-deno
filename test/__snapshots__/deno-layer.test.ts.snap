// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DenoLayer should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DenoLayerD97F84F7": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f8449f418ec9d05be251070898cc7989f8dc1ca8cc571cbcf51bdc46f719c375.zip",
        },
        "Description": "/bin/deno",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
